### ISSUE TITLE
[MENFORCER-195] Dependency convergence does not support wildcard exclusions

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-enforcer#692]
**[Tomaz Cerar](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ctomc)** opened **[MENFORCER-195](https://issues.apache.org/jira/browse/MENFORCER-195?redirect=false)** and commented

Maven 3.2 introduced wildcard exclusions for dependencies.

But if you use them dependencyConvergence wrongly complains about conflicting transitive dependencies.

**WORKAROUND Maven 3.6.2:** Start the exclusion-list with the explicit groupID and artifactId, ending with the exclusions containing wildcards. This will restore the behavior as before MNG-6713

---

**Affects:** 1.3.1

**Issue Links:**
- [MSHARED-816](https://issues.apache.org/jira/browse/MSHARED-816) Annotate DependencyNodes with dependency management metadata
 (_**"is blocked by"**_)
- [MENFORCER-232](https://issues.apache.org/jira/browse/MENFORCER-232) DependencyConvergence doesn't acknowledge exclusion wildcards (MNG-2315)
 (_**"is duplicated by"**_)
- [MENFORCER-302](https://issues.apache.org/jira/browse/MENFORCER-302) DependencyConvergence does not recognize exclusion wildcards
 (_**"is duplicated by"**_)
- [MDEP-374](https://issues.apache.org/jira/browse/MDEP-374) dependency:tree -Dverbose isn't verbose anymore

- [MNG-6713](https://issues.apache.org/jira/browse/MNG-6713) Fix ExclusionArtifactFilter to respect wildcard
 (_**"requires"**_)
- [MSHARED-339](https://issues.apache.org/jira/browse/MSHARED-339) DependencyGraphBuilder does not provide verbose tree
 (_**"depends upon"**_)

**Remote Links:**
- [GitHub Pull Request #20
](https://github.com/apache/maven-enforcer/pull/20)
- [GitHub Pull Request #54
](https://github.com/apache/maven-enforcer/pull/54)
- [GitHub Pull Request #55
](https://github.com/apache/maven-enforcer/pull/55)
- [GitHub Pull Request #56
](https://github.com/apache/maven-enforcer/pull/56)

12 votes, 21 watchers

**Comment by jira-importer** [comment]
**[Tomaz Cerar](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ctomc)** commented

example of dependancy definition:

```xml
  <dependency>
                <groupId>org.apache.myfaces.core</groupId>
                <artifactId>myfaces-impl</artifactId>
                <version>2.1.8</version>
                <exclusions>
                    <exclusion>
                        <groupId>*</groupId>
                        <artifactId>*</artifactId>
                    </exclusion>
          </exclusions>
</dependency>
```

**Comment by jira-importer** [comment]
**[Paul Gier](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=pgier)** commented

I wasn't able to reproduce a build failure.  I experimented with enforcer 1.3, 1.3.1, and the latest 2.0-SNAPSHOT.  Using Maven 3.1 and lower I saw the following warnings:

```
[WARNING] 
[WARNING] Some problems were encountered while building the effective model for org.apache.maven.its.enforcer:ban-transitive-test:jar:1.0
[WARNING] 'dependencies.dependency.exclusions.exclusion.groupId' for org.apache.myfaces.core:myfaces-impl:jar with value '*' does not match a valid id pattern. @ line 51, column 20
[WARNING] 'dependencies.dependency.exclusions.exclusion.artifactId' for org.apache.myfaces.core:myfaces-impl:jar with value '*' does not match a valid id pattern. @ line 52, column 23
[WARNING] 
```

But the dependencies appeared to be excluded, and the build was successful.
Using Maven 3.2.1, there were no warnings, and the dependencies were excluded as expected.

**Comment by jira-importer** [comment]
**[Tomaz Cerar](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ctomc)** commented

I can reproduce it also with 2.0-SNAPSHOT

```
[INFO] --- maven-enforcer-plugin:2.0-SNAPSHOT:enforce (ban-bad-dependencies) @ wildfly-build ---
[WARNING]
Dependency convergence error for commons-httpclient:commons-httpclient:3.1 paths to dependency are:
+-org.wildfly:wildfly-build:9.0.0.Alpha1-SNAPSHOT
  +-org.opensaml:openws:1.4.4
    +-commons-httpclient:commons-httpclient:3.1
and
+-org.wildfly:wildfly-build:9.0.0.Alpha1-SNAPSHOT
  +-org.opensaml:xmltooling:1.3.4
    +-ca.juliusdavies:not-yet-commons-ssl:0.3.9
      +-commons-httpclient:commons-httpclient:3.0
and
+-org.wildfly:wildfly-build:9.0.0.Alpha1-SNAPSHOT
  +-org.jboss.narayana:jbosstxbridge:5.0.2.Final
    +-commons-httpclient:commons-httpclient:3.1-jbossorg-1
```

where org.jboss.narayana:jbosstxbridge and org.opensaml:xmltooling:1.3.4 are defined as:

```xml
<dependency>
                <groupId>org.jboss.narayana</groupId>
                <artifactId>jbosstxbridge</artifactId>
                <version>${version.org.jboss.narayana}</version>
                <exclusions>
                    <exclusion>
                        <groupId>*</groupId>
                        <artifactId>*</artifactId>
                    </exclusion>
                </exclusions>
            </dependency>
<dependency>
                <groupId>org.opensaml</groupId>
                <artifactId>xmltooling</artifactId>
                <version>${version.org.opensaml.xmltooling}</version>
                <exclusions>
                    <exclusion>
                        <groupId>*</groupId>
                        <artifactId>*</artifactId>
                    </exclusion>
                </exclusions>
            </dependency>

```

**Comment by jira-importer** [comment]
**[Paul Gier](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=pgier)** commented

Ok, the problem is in the dependencyConvergence rule, I was looking at the banTransitiveDependencies rule.  I can reproduce it now.

**Comment by jira-importer** [comment]
**[Henning Hoefer](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hoeferh)** commented

> This is a PR merged from a forked repository.

Contrary to the bot's message, this PR has **not** been merged! (but the PR has been closed)

The problem persists, still affects 3.0.0.M2

**Comment by jira-importer** [comment]
**[Paul Vonnahme](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=pmv)** commented

It looks like the comment in https://issues.apache.org/jira/browse/MSHARED-339?focusedCommentId=16817496&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16817496 would work to fix the issue.  However, it requires adopting the 

```
DependencyNode buildDependencyGraph( ProjectBuildingRequest buildingRequest, ArtifactFilter filter )
    throws DependencyGraphBuilderException;
```

method signature.  The current version of maven-dependency-tree that the enforcer uses has

```
DependencyNode buildDependencyGraph(MavenProject var1, ArtifactFilter var2) throws DependencyGraphBuilderException;
```

I don't think that is a problem assuming the enforcer plugin is willing to update that dependency. Ideally this could be included in the final 3.0.0 version of the enforcer.

I started on the changes and hoped to open a PR, but fixing DependencyConvergence breaks RequireUpperBoundDeps - a rule I'm not all that familiar with.  It seems DependencyNode.getPremanagedVersion() is null, while with DependencyTreeBuilder it had a value. After more research, it seems this isn't implemented, and MSHARED-816 needs done for that to work.

To recap:
1. To fix the dependency convergence rule (this ticket), the maven-dependency-tree dependency needs updated
2. Once the maven-dependency-tree dependency is updated, all rules that used dependencies need changed since DependencyTreeBuilder does not exist in the new version
3. Once everything is converted to DependencyGraphBuilder, "RequireUpperBoundDeps" starts failing since it requires MSHARED-816

**Comment by jira-importer** [comment]
**[Fabian Meier](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=fabianmeier)** commented

Sehr geehrte Damen und Herren,

ich bin ab 3. Juni wieder im Haus.

Viele Grüße
Fabian Meier (ik3-mv)

**Comment by jira-importer** [comment]
**[Paul Vonnahme](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=pmv)** commented

In addition, if a new version of the enforcer is released with the new maven-dependency-tree dependency, I believe extra-enforcer-rules will have to have a version released at the same time to be compatible, since it also references DependencyTreeBuilder: https://github.com/mojohaus/extra-enforcer-rules/search?q=DependencyTreeBuilder&unscoped_q=DependencyTreeBuilder

**Comment by jira-importer** [comment]
**[Thomas Mortagne](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tmortagne)** commented

Thanks for the fast answer Robert. Let's wait for 3.6.3 then :)

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

[Thomas Mortagne](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tmortagne) I would expect that the following workaround works: move all explicit exclusions to the top, wildcards to the bottom. Can you confirm?
